### PR TITLE
Make header decoding fail when MAX_FIELD_SECTION_SIZE is exceeded

### PR DESCRIPTION
**Pull request by Ruben2424**
Hi,

I solved #53. I am new to Rust, so every feedback is appreciated.

### CODE REVIEW

File: h3/src/client.rs [hunk 1]
```diff
@@ -288,27 +288,29 @@ where
             .ok_or_else(|| {
                 Code::H3_GENERAL_PROTOCOL_ERROR.with_reason("Did not receive response headers")
             })?;
-
-        let decoded = if let Frame::Headers(ref mut encoded) = frame {
-            qpack::decode_stateless(encoded)?
+        
+        let decoded;
```
Comment (seanmonstar) [inline h3/src/client.rs:292]: To make this lint happy, you can write `let decoded = if let Frame::Headers(..)`.

File: h3/src/connection.rs [hunk 2]
```diff
@@ -340,6 +340,7 @@ where
 
     /// Receive trailers
     pub async fn recv_trailers(&mut self) -> Result<Option<HeaderMap>, Error> {
+
```
Comment (seanmonstar) [inline h3/src/connection.rs:343]:
```suggestion
```

File: h3/src/connection.rs [hunk 3]
```diff
@@ -366,11 +367,12 @@ where
         }
 
         let qpack::Decoded {
-            fields, mem_size, ..
-        } = qpack::decode_stateless(&mut trailers)?;
-        if mem_size > self.max_field_section_size {
-            return Err(Error::header_too_big(mem_size, self.max_field_section_size));
-        }
+            fields, ..
+        } =  match qpack::decode_stateless(&mut trailers,self.max_field_section_size){
```
Comment (seanmonstar) [inline h3/src/connection.rs:371]:
```suggestion
        } = match qpack::decode_stateless(&mut trailers,self.max_field_section_size) {
```

File: h3/src/qpack/decoder.rs [hunk 4]
```diff
@@ -35,6 +35,7 @@ pub enum Error {
     MissingRefs(usize),
     BadBaseIndex(isize),
     UnexpectedEnd,
+    HeaderToLong(u64),
```
Comment (seanmonstar) [inline h3/src/qpack/decoder.rs:38]:
```suggestion
    HeaderTooLong(u64),
```

File: h3/src/qpack/decoder.rs [hunk 5]
```diff
@@ -51,6 +52,7 @@ impl std::fmt::Display for Error {
             Error::MissingRefs(n) => write!(f, "missing {} refs to decode bloc", n),
             Error::BadBaseIndex(i) => write!(f, "out of bounds base index: {}", i),
             Error::UnexpectedEnd => write!(f, "unexpected end"),
+            Error::HeaderToLong(_) => write!(f, "header to long "),
```
Comment (seanmonstar) [inline h3/src/qpack/decoder.rs:55]:
```suggestion
            Error::HeaderTooLong(_) => write!(f, "header too long"),
```

File: h3/src/qpack/decoder.rs [hunk 6]
```diff
@@ -235,6 +237,10 @@ pub fn decode_stateless<T: Buf>(buf: &mut T) -> Result<Decoded, Error> {
             _ => return Err(Error::UnknownPrefix(buf.chunk()[0])),
         };
         mem_size += field.mem_size() as u64;
+        // Cancel decoding if the header is considered too big
+        if mem_size > max_size{
```
Comment (seanmonstar) [inline h3/src/qpack/decoder.rs:241]:
```suggestion
        if mem_size > max_size {
```

**Review by seanmonstar — CHANGES_REQUESTED** [review]
Thanks for the PR! I left some style comments inline. I'm also curious, is there a test case we could add to verify the change?

File: h3/src/client.rs (hunk 1 above)
Comment (Ruben2424) [reply h3/src/client.rs:292]: I tried this, but then `decoded` has the type of `()` which leads to a type mismatch in this line `let qpack::Decoded { fields, .. } = decoded;`. This is the code:

```rust
        let decoded = if let Frame::Headers(ref mut encoded) = frame {
            match qpack::decode_stateless(encoded, self.inner.max_field_section_size) {
                Err(qpack::DecoderError::HeaderTooLong(cancel_size)) => {
                    self.inner.stop_sending(Code::H3_REQUEST_CANCELLED);
                    return Err(Error::header_too_big(
                        cancel_size,
                        self.inner.max_field_section_size,
                    ));
                }
                Ok(decoded) => decoded,
                Err(e) => return Err(e.into()),
            };
        } else {
            return Err(
                Code::H3_FRAME_UNEXPECTED.with_reason("First response frame is not headers")
            );
        };

        let qpack::Decoded { fields, .. } = decoded;
```
What did i wrong?

**Comment by Ruben2424** [comment]
I fixed the style issues and added a test. I actually didn't know the cargo fmt tool.

File: h3/src/client.rs (hunk 1 above)
Comment (seanmonstar) [reply h3/src/client.rs:292]: I think you just need to the remove the semi-colon from the `match` expression.

Comment (Ruben2424) [reply h3/src/client.rs]: Ah, that helped. Thank you.

**Review by seanmonstar — APPROVED** [review]
Looks good here, thanks!

**Pull request merged**